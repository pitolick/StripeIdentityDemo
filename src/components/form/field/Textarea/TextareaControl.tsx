/**
 * テキストエリア制御コンポーネント（react-hook-form）
 * @module TextareaControl
 */
import React, { forwardRef } from 'react'
import { Textarea, TextareaProps } from './Textarea'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * テキストエリアコンポーネントのプロパティ
 * @memo バリデーションスキーマ（zod）は各入力子コンポーネント内でexportしformコンポーネント内でresolverに指定する
 * @typedef {Object} TextareaControlProps
 */
export interface TextareaControlProps extends TextareaProps, Omit<InputBoxProps, 'children'> {}

/**
 * テキストエリア制御コンポーネント（react-hook-form）
 * @param {TextareaControlProps} props - テキストエリアコンポーネントのプロパティ
 * @returns {JSX.Element} - テキストエリアコンポーネント
 */
export const TextareaControl = forwardRef<HTMLTextAreaElement, TextareaControlProps>((props, ref) => {
  const { nameId, label, note, required = undefined, className, error, contentWidth, ...textareaProps } = props
  const classNameProps = className ? className : ''
  // エラー時のclass名
  const errorClass = error ? `bg-red-pale outline outline-red outline-2 -outline-offset-2` : ''

  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      {/* テキストエリア */}
      <Textarea id={nameId} ref={ref} {...textareaProps} className={`${errorClass} ${classNameProps}`} />
    </InputBox>
  )
})

TextareaControl.displayName = 'TextareaControl'
