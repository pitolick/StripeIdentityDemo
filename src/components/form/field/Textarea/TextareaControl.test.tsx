import React from 'react'
import { render } from '@testing-library/react'
import { TextareaControl, TextareaControlProps } from './TextareaControl'

describe('TextareaControl', () => {
  it('正しくレンダリングされること', () => {
    const defaultProps: TextareaControlProps = {
      nameId: 'textarea',
      label: 'テキストエリア',
      note: { list: ['テキストエリアの説明'] },
      required: true,
      className: 'custom-class',
      error: { type: 'test' },
      contentWidth: 'full',
    }

    const { container } = render(<TextareaControl {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  it('デフォルト値が正しく設定されること', () => {
    const defaultProps: TextareaControlProps = {
      nameId: 'textarea',
    }

    const { container } = render(<TextareaControl {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })
})
