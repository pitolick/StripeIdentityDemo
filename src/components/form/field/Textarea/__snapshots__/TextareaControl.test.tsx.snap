// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaControl デフォルト値が正しく設定されること 1`] = `
<div>
  <div
    class="false space-y-2 false"
  >
    <div
      class="space-y-2 false"
    >
      <div
        class="flex w-full items-center justify-start gap-x-3"
      >
        <textarea
          class="  block w-full rounded-md border border-gray-500 px-3 py-2 text-sm md:px-5  md:py-3 md:text-base"
          id="textarea"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TextareaControl 正しくレンダリングされること 1`] = `
<div>
  <div
    class="w-full flex-auto space-y-2 false"
  >
    <div
      class="flex items-center justify-start gap-x-2"
    >
      <span
        class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
      >
        必須
      </span>
      <label
        class="block text-sm font-bold"
        for="textarea"
      >
        テキストエリア
      </label>
    </div>
    <div
      class="space-y-2 false"
    >
      <div
        class="flex w-full items-center justify-start gap-x-3"
      >
        <textarea
          class="bg-red-pale outline outline-red outline-2 -outline-offset-2 custom-class block w-full rounded-md border border-gray-500 px-3 py-2 text-sm md:px-5  md:py-3 md:text-base"
          id="textarea"
        />
      </div>
      <ul>
        <li
          class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
        >
          テキストエリアの説明
        </li>
      </ul>
      <p
        class="text-sm text-red"
      />
    </div>
  </div>
</div>
`;
